File: 2023-2024/sourcecode/mod4/lab12/ResourcePatternBasicsJava/app/src/main/java/com/example/resourcepatternbasicsjava/viewmodel/CommentsViewModel.java
package com.example.resourcepatternbasicsjava.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.resourcepatternbasicsjava.data.CommentResponseItem;
import com.example.resourcepatternbasicsjava.data.repository.CommentRepository;
import com.example.resourcepatternbasicsjava.util.Resource;

import java.io.IOException;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CommentsViewModel extends ViewModel {

    private final CommentRepository repository = new CommentRepository();
    private final MutableLiveData<Resource<List<CommentResponseItem>>> _comments = new MutableLiveData<>();
    public LiveData<Resource<List<CommentResponseItem>>> comments = _comments;

    public CommentsViewModel() {
        try {
            getPosts();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void getPosts() throws IOException {
        Call<List<CommentResponseItem>> response = repository.getComments();

        response.enqueue(new Callback<List<CommentResponseItem>>() {
            @Override
            public void onResponse(@NonNull Call<List<CommentResponseItem>> call, @NonNull Response<List<CommentResponseItem>> response) {
                if (response.isSuccessful()){
                    _comments.postValue(Resource.loading(null));
                    List<CommentResponseItem> postsResponse = response.body();
                    if (postsResponse != null)
                        _comments.postValue(Resource.success(postsResponse));
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<CommentResponseItem>> call, @NonNull Throwable t) {
                _comments.postValue(Resource.error(t.getLocalizedMessage(), null));
            }
        });
    }
}

File: 2023-2024/sourcecode/mod4/lab12/ResourcePatternBasicsKotlin/app/src/main/java/com/example/resourcepatternbasicskotlin/viewmodel/CommentsViewModel.kt
package com.example.resourcepatternbasicskotlin.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.resourcepatternbasicskotlin.data.CommentResponseItem
import com.example.resourcepatternbasicskotlin.data.repository.CommentRepository
import com.example.resourcepatternbasicskotlin.util.Resource
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import retrofit2.Response

class CommentsViewModel : ViewModel() {
    private val repository = CommentRepository()
    private var _comments: MutableStateFlow<Resource<List<CommentResponseItem>>> = MutableStateFlow(Resource.Loading())
    val comments: StateFlow<Resource<List<CommentResponseItem>>> = _comments

    init {
        getCommentsList()
    }

    private fun getCommentsList() = viewModelScope.launch {
        _comments.value = Resource.Loading()
        val response = repository.getComments()
        _comments.value = handleCommentsResponse(response)
    }

    private fun handleCommentsResponse(response: Response<List<CommentResponseItem>>)
            : Resource<List<CommentResponseItem>> {
        if (response.isSuccessful)
            response.body()?.let { return Resource.Success(it) }
        return Resource.Error(response.message())
    }
}